### PR TITLE
[MI-2039]: Publishing Service Now VA is typing until we get response from the bot 

### PR DESCRIPTION
**Pull request by Nityanand13**

### CODE REVIEW

File: server/plugin/api_test.go [hunk 1]
```diff
@@ -862,13 +862,27 @@ func Test_handleDateTimeSelection(t *testing.T) {
...
 	} {
 		t.Run(name, func(t *testing.T) {
+			mockInterval := int64(1000)
 			p := new(Plugin)
 
 			mockAPI := &plugintest.API{}
 			mockAPI.On("GetBundlePath").Return("mockString", nil)
 			mockAPI.On("LogDebug", testutils.GetMockArgumentsWithType("string", 7)...).Return("LogDebug error")
-			mockAPI.On("LogError", testutils.GetMockArgumentsWithType("string", 6)...).Return("LogError error")
+			mockAPI.On("LogError", mock.Anything, mock.Anything, mock.Anything, mock.Anything, mock.Anything).Return("LogError error")
```
Comment (raghavaggarwal2308) [inline server/plugin/api_test.go]: why this change?

File: server/plugin/api_test.go [hunk 2]
```diff
@@ -862,13 +862,27 @@ func Test_handleDateTimeSelection(t *testing.T) {
...
+					TimeBetweenUserTypingUpdatesMilliseconds: &mockInterval,
+				},
+			})
+
+			mockAPI.On("GetDirectChannel", mock.Anything, mock.Anything).Return(&model.Channel{
+				Id: "mock-channelID",
+			}, nil)
+
+			mockAPI.On("KVSetWithOptions", mock.Anything, mock.Anything, mock.Anything).Return(true, nil)
+			mockAPI.On("KVGet", mock.Anything).Return([]byte{}, nil)
```
Comment (raghavaggarwal2308) [inline server/plugin/api_test.go]: We are returning nil in all these cases that means we are not covering the code completely

File: server/plugin/plugin.go [hunk 3]
```diff
@@ -45,6 +48,52 @@ func (p *Plugin) OnActivate() error {
...
+func (p *Plugin) closeBackgroundJob() {
+	if err := p.backgroundJob.Close(); err != nil {
+		p.API.LogError("Failed to close background job", "Error", err.Error())
+	}
+}
+
+func (p *Plugin) deactivateJob() {
+	if p.backgroundJob != nil {
+		p.closeBackgroundJob()
+		if err := p.API.KVDelete("cron_" + PublishSeriveNowVAIsTypingJobName); err != nil {
```
Comment (raghavaggarwal2308) [inline server/plugin/plugin.go]: This is not a cron job so we should use different key prefix

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

**Comment by manojmalik20** [comment]
Change the base branch to something else like `loader_support`.

File: server/plugin/api_test.go (hunk 1 above)
Comment (Nityanand13) [reply server/plugin/api_test.go]: Because now we have different types of error

File: server/plugin/api_test.go (hunk 2 above)
Comment (Nityanand13) [reply server/plugin/api_test.go]: Lets discuss this on a call

**Review by raghavaggarwal2308 — APPROVED** [review]

File: server/plugin/hooks.go [hunk 4]
```diff
@@ -83,6 +83,11 @@ func (p *Plugin) MessageHasBeenPosted(c *plugin.Context, post *model.Post) {
 		}
 	}
 
+	if err = p.scheduleJob(mattermostUserID); err != nil {
+		p.API.LogError("Error while scheduling a job", "Error", err.Error())
```
Comment (manojmalik20) [inline server/plugin/hooks.go]: Instead of logging the error everywhere where you're calling `scheduleJob`, you can just log the error inside `scheduleJob`.

File: server/plugin/plugin.go [hunk 5]
```diff
@@ -56,6 +59,52 @@ func (p *Plugin) OnDeactivate() error {
...
+func (p *Plugin) closeBackgroundJob() {
+	if err := p.backgroundJob.Close(); err != nil {
+		p.API.LogError("Failed to close background job", "Error", err.Error())
+	}
+}
+
+func (p *Plugin) deactivateJob() {
+	if p.backgroundJob != nil {
+		p.closeBackgroundJob()
+		if err := p.API.KVDelete(cronPrefix + PublishSeriveNowVAIsTypingJobName); err != nil {
```
Comment (manojmalik20) [inline server/plugin/plugin.go:71]: No need for the cron prefix. Instead, you should hash the key. See https://github.com/Brightscout/mattermost-plugin-hamilton-artifactory/blob/063b96222d3fcbc16506964b31a8ce95b6618e30/server/plugin.go#L91

File: server/plugin/plugin.go [hunk 6]
```diff
@@ -56,6 +59,52 @@ func (p *Plugin) OnDeactivate() error {
...
+	}
+
+	intervalInSecond := time.Duration(interval) * time.Second
+	job, cronErr := cluster.Schedule(
+		p.API,
+		PublishSeriveNowVAIsTypingJobName,
+		cluster.MakeWaitForRoundedInterval(intervalInSecond),
+		func() {
+			if err = p.API.PublishUserTyping(p.botUserID, channel.Id, ""); err != nil {
+				p.API.LogError("Failed to publish a user is typing WebSocket event", "Error", err.Error())
```
Comment (manojmalik20) [inline server/plugin/plugin.go]:
```suggestion
				p.API.LogDebug("Failed to publish a user is typing WebSocket event", "Error", err.Error())
```

File: server/plugin/api_test.go [hunk 7]
```diff
@@ -860,15 +863,57 @@ func Test_handleDateTimeSelection(t *testing.T) {
...
 			p := new(Plugin)
 
 			mockAPI := &plugintest.API{}
 			mockAPI.On("GetBundlePath").Return("mockString", nil)
 			mockAPI.On("LogDebug", testutils.GetMockArgumentsWithType("string", 7)...).Return("LogDebug error")
-			mockAPI.On("LogError", testutils.GetMockArgumentsWithType("string", 6)...).Return("LogError error")
+			mockAPI.On("LogError", testutils.GetMockArgumentsWithType("string", 5)...).Return()
 			mockAPI.On("UpdatePost", mock.AnythingOfType("*model.Post")).Return(nil, nil)
+
+			mockAPI.On("GetConfig").Return(&model.Config{
```
Comment (manojmalik20) [inline server/plugin/api_test.go:889]: Instead of doing all this, use monkey patch and mock the return type of `p.scheduleJob`.

File: server/plugin/hooks_test.go [hunk 8]
```diff
@@ -110,6 +126,25 @@ func Test_MessageHasBeenPosted(t *testing.T) {
 				mockAPI.On("LogDebug", testutils.GetMockArgumentsWithType("string", 3)...).Return()
 			}
 
+			if !(testCase.getChannelError != nil || testCase.getUserError != nil || testCase.createMessageAttachmentError != nil || testCase.parseAuthTokenError != nil || testCase.Message == DisconnectKeyword) {
```
Comment (manojmalik20) [inline server/plugin/hooks_test.go]: Use monkey patch to mock `p.scheduleJob`

File: server/plugin/user_test.go [hunk 9]
```diff
@@ -379,6 +399,24 @@ func Test_CompleteOAuth2(t *testing.T) {
...
+				ServiceSettings: model.ServiceSettings{
+					TimeBetweenUserTypingUpdatesMilliseconds: &mockInterval,
+				},
+			})
+
+			mockAPI.On("GetDirectChannel", mock.Anything, mock.Anything).Return(&model.Channel{
+				Id: "mock-channelID",
+			}, testCase.getDirectChannelError)
+
+			monkey.Patch(cluster.Schedule, func(_ cluster.JobPluginAPI, _ string, _ cluster.NextWaitInterval, _ func()) (*cluster.Job, error) {
```
Comment (manojmalik20) [inline server/plugin/user_test.go]: Same here.

File: server/plugin/plugin.go [hunk 10]
```diff
@@ -87,17 +87,20 @@ func (p *Plugin) scheduleJob(mattermostUserID string) error {
 	}
 
 	intervalInSecond := time.Duration(interval) * time.Second
+
+	// Schedule creates a scheduled job and stores the key in kv store
```
Comment (manojmalik20) [inline server/plugin/plugin.go]:
```suggestion
	// cluster.Schedule creates a scheduled job and stores job metadata in kv store using key "cron_<jobName>"
```

**Review by manojmalik20 — APPROVED** [review]

**Review by manojmalik20 — APPROVED** [review]
